### PR TITLE
[feature] Adds fusion techniques

### PR DESCRIPTION
**Pull request by pushkalkatara**
This PR solves issue #162 .

### CODE REVIEW

**Comment by pushkalkatara** [comment]
Hi @apsdehal I have added all the fusions techniques from the remi's repo. Also, the CompactBilinearPooling was implemented using a package pytorch_fft, but recently there was fft support in pytorch. I changed the CompactBilinearPooling class accordingly and merged all in one file `fusions.py`. Wrote some tests also to check all the fusion techniques in `test_fusions.py`

**Comment by pushkalkatara** [comment]
Oh, I'll make the changes soon.
Remi's repo is also dependent on the package [pytorch_fft](https://github.com/locuslab/pytorch_fft) which was developed for fft when it was not part of pytorch. but recently [fft](https://pytorch.org/cppdocs/api/function_namespaceat_1a3c00c632a6391e6fb8246938ad254da1.html) was implemented in pytorch.

**Comment by pushkalkatara** [comment]
@apsdehal I made the changes. Only fusion.MCB fails as it requires pytorch_fft package, requirements are listed separately [here](https://github.com/Cadene/block.bootstrap.pytorch/blob/master/requirements_mcb.txt) but not in the package. It was developed using pytorch_fft package when fft was not implemented in pytorch but recently it was [added](https://pytorch.org/cppdocs/api/function_namespaceat_1a3c00c632a6391e6fb8246938ad254da1.html). Should I add these in our requirements file or update the code with pytorch/fft?

**Comment by apsdehal** [comment]
Update the fusion technique which uses pytorch/fft to use our own version. Build is failing at this moment.

**Comment by pushkalkatara** [comment]
Hi I have made a PR to remi's repo - [PR](https://github.com/Cadene/block.bootstrap.pytorch/pull/15), the build fails as pytorch_fft is not added as a repository. As soon as the above PR is merged, the build would complete.

**Comment by apsdehal** [comment]
Thank you for making that PR. I will wait till that one merges. If it takes more than a week, I can merge this one and fix the build.
Feel free to work on other issues if you please. Do ask me before picking up one of them or if you want something new I can come with important stuff pythia needs too. :) Thanks for working on Pythia. :)

**Comment by pushkalkatara** [comment]
With the new version of block.bootstrap.pytorch the build completes.

File: pythia/modules/fusions.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import collections
+
+import torch
+import torch.nn as nn
+from block import fusions
+
+from pythia.common.registry import registry
+
+
+@registry.register_fusion("block")
```
Comment (apsdehal) [inline pythia/modules/fusions.py]: Instead of doing this, you can just do 

`registry.register_fusion("block")(fusion.Block)` not the no `@` here.

File: pythia/modules/fusions.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    from torch import nn
+
+    @regitery.register_fusion("custom")
+    class CustomFusion(nn.Module):
+        def __init__(self, params=None):
+            super().__init__("Custom")
+
+Then in your model's config you can specify ``fusion`` attribute as follows:
+
+.. code::
```
Comment (apsdehal) [inline pythia/modules/fusions.py]: I don't this will automatically create a fusion layer inside the model as there is no fusion base class. So, let's just remove this line 19-27 for now.

File: pythia/modules/fusions.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+"""
+The fusions module contains various Fusions techniques some based on BLOCK:
+Bilinear Superdiagonal Fusion for VQA and VRD. For e.g. LinearSum, ConcatMLP
+taken from https://github.com/Cadene/block.bootstrap.pytorch#fusions.
+
+For implementing your own fusion technique, you need to follow these steps:
+
+.. code::
+    from pythia.common.registry import registry
+    from pythia.modules.fusions import BaseFusion
```
Comment (apsdehal) [inline pythia/modules/fusions.py]: We don't have a `BaseFusion` class anymore.
Can we list all of the Fusions available here?

File: tests/modules/test_fusions.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    def test_ConcatMLP(self):
+        fusion = fusions.ConcatMLP(self.input_dims, self.output_dims, dimensions=[5,5])
+        out = fusion(self.x)
+        if torch.cuda.is_available():
+            fusion.cuda()
+            out = fusion([self.x[0].cuda(), self.x[1].cuda()])
+        assert torch.Size([2,2]) == out.shape
+
+if __name__ == '__main__':
```
Comment (apsdehal) [inline tests/modules/test_fusions.py]: Thanks for writing these great tests. Though these tests don't test the working of the fusion layer itself. You should set a seed in start and then compare the output values to pre-saved values for results to make sure something hasn't broken.

Comment (pushkalkatara) [reply tests/modules/test_fusions.py]: Do I need to check the output as well by setting manual seed and assert almost equal with the hardcoded values as in the [layer tests](https://github.com/facebookresearch/pythia/blob/2ab8083f3da6a5421c5f5312df3b4ca7cccb230c/tests/modules/test_layers.py#L25)

Comment (apsdehal) [reply tests/modules/test_fusions.py]: Yes, that should be case. Let's fix that in a separate PR.

**Comment by pushkalkatara** [comment]
Thanks for the merge :D

**Comment by apsdehal** [comment]
Thank you for working on this and making Pythia better! This has now landed into master. Open up a separate PR for test fixes whenever you get time. :)
